utils/tutorial_steps.ts

```diff
@@ -0,0 +1,43 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import {createSelector} from 'reselect';
+
+import {getCurrentUser} from 'mattermost-redux/selectors/entities/users';
+import {GlobalState} from 'mattermost-redux/types/store';
+
+export type TSteps = Record<string, number>;
+
+export const AdminTutorialSteps = {
+    ADD_FIRST_CHANNEL: -1,
+    POST_POPOVER: 0,
+    CHANNEL_POPOVER: 1,
+    ADD_CHANNEL_POPOVER: 2,
+    MENU_POPOVER: 3,
+    PRODUCT_SWITCHER: 4,
+    SETTINGS: 5,
+    START_TRIAL: 6,
+    FINISHED: 999,
+};
+
+export const TutorialSteps = {
+    ADD_FIRST_CHANNEL: -1,
+    POST_POPOVER: 0,
+    CHANNEL_POPOVER: 1,
+    ADD_CHANNEL_POPOVER: 2,
+    MENU_POPOVER: 3,
+    PRODUCT_SWITCHER: 4,
+    SETTINGS: 5,
+    FINISHED: 999,
+};
+
+export const getTutorialSteps = createSelector(
```

Comment:
I like the intent of what this is doing of declaratively controlling which tips the user sees, but I don't think we can safely do it until we migrate the data so that each tip is its own independent piece of data in the database.

So for now, I'd 3/5 recommend having a single set of tutorial steps instead of splitting what the numbers mean per user/role. To avoid non-admins seeing the start trial tip, we can instead control seeing or skipping the tip in whichever component is most relevant to control it from. We have a pattern for doing this in the product switcher tip: https://github.com/mattermost/mattermost-webapp/blob/master/components/global_header/left_controls/product_menu/product_menu_tip/product_menu_tip.tsx#L60-L90

Some reasons to avoid splitting the meaning of the step preference number until we migrate the data:
* User role is mutable: non-admins can become admins, admins can stop being admins. When they do, their tip preference data will be interpreted incorrectly
* When we add on new tips, we need to maintain these two lists
* If other tips become conditional on other criteria than admin status, we need even more lists
* If we go the split list route today, it make migrating more complicated in the future because the migration also needs to take admin status into account, and any users whose admin role changed in the meantime will be migrated incorrectly.

The long term solution (probably outside the scope of this ticket and PR) is to migrate this data out of a single user preference into something that supports each tip's viewed status as its own prop, and we control view and ordering from the UI (like is being done here). We could use a boolean preference per tip, or maybe we could store tips in the sql `users.props` jsonb column.

Reply:
@neallred agreed. Thanks for the detailed explanation. I am going to make edits